File: solutions/ruby/two-bucket/2/two_bucket.rb
class TwoBucket
  BUCKET_ONE = "one"
  BUCKET_TWO = "two"

  attr_reader :moves, :goal_bucket, :other_bucket

  def initialize(small_size, large_size, goal, starting_bucket)
    @moves        = 0
    @goal_bucket  = ""
    @other_bucket = 0 # Amount of non-goal bucket

    starts_with_one = starting_bucket == BUCKET_ONE
    non_starting_bucket = starts_with_one ? BUCKET_TWO : BUCKET_ONE

    first_bucket  = Bucket.new(starts_with_one ? small_size : large_size)
    second_bucket = Bucket.new(starts_with_one ? large_size : small_size)

    while @goal_bucket.empty?
      @moves += 1

      if @moves == 1
        first_bucket.fill # Must fill first bucket on first move
      elsif @moves == 2 && second_bucket.capacity == goal
        second_bucket.fill # Edge case shortcut to reach goal
      elsif @moves.even?
        first_bucket.pour_into(second_bucket)
      elsif second_bucket.full?
        second_bucket.drain
      else
        first_bucket.fill
      end

      @goal_bucket = starting_bucket     if first_bucket.amount  == goal
      @goal_bucket = non_starting_bucket if second_bucket.amount == goal
    end

    @other_bucket = first_bucket.amount  if @goal_bucket == non_starting_bucket
    @other_bucket = second_bucket.amount if @goal_bucket == starting_bucket

  end
end

class Bucket
  attr_reader :capacity, :amount

  def initialize(capacity)
    @capacity = capacity
    @amount   = 0
  end

  def fill(amount = free_amount)
    @amount += amount
  end

  def drain(amount = @amount)
    @amount -= amount
  end

  def pour_into(bucket)
    displacement = @amount >= bucket.free_amount ? bucket.free_amount : @amount

    bucket.fill(displacement)
    drain(displacement)
  end

  def free_amount
    @capacity - @amount
  end

  def full?
    @amount == @capacity
  end
end

File: solutions/javascript/two-bucket/1/two-bucket.js
class Bucket {
  constructor(capacity) {
    this.capacity = capacity;
    this.amount   = 0;
  }

  fill(amount) {
    if (typeof amount === 'undefined') amount = this.capacity;
    this.amount += amount;
  }

  drain(amount) {
    if (typeof amount === 'undefined') amount = this.amount;
    this.amount -= amount;
  }

  freeAmount() {
    return this.capacity - this.amount;
  }

  pourInto(bucket) {
    const displacement = this.amount >= bucket.freeAmount()
      ? bucket.freeAmount()
      : this.amount;

    bucket.fill(displacement);
    this.drain(displacement);
  }

  isFull() {
    return this.amount === this.capacity;
  }
}

export class TwoBucket {
  #solution = { moves: 0, goalBucket: '', otherBucket: 0 };

  constructor(smallSize, largeSize, goal, startingBucket) {
    const bucketOne         = 'one';
    const bucketTwo         = 'two';
    const startsWithOne     = startingBucket === bucketOne;
    const nonStartingBucket = startsWithOne ? bucketTwo : bucketOne;

    let firstBucket  = new Bucket(startsWithOne ? smallSize : largeSize);
    let secondBucket = new Bucket(startsWithOne ? largeSize : smallSize);
    let states       = [];

    while (!this.#solution.goalBucket) {
      const state = [firstBucket.amount, secondBucket.amount];

      if (states.some(s => s[0] === state[0] && s[1] === state[1])) {
        throw new Error('Unreachable goal.');
      }

      states.push(state);
      this.#solution.moves++;

      if (this.#solution.moves === 1) {
        firstBucket.fill(); // Must fill first bucket on first move.
      }
      else if (this.#solution.moves === 2 && secondBucket.capacity === goal) {
        secondBucket.fill(); // Edge case shortcut to reach goal on second move.
      }
      else if (this.#solution.moves % 2 === 0) {
        firstBucket.pourInto(secondBucket);
      }
      else if (secondBucket.isFull()) {
        secondBucket.drain();
      }
      else {
        firstBucket.fill();
      }

      if (firstBucket.amount === goal) {
        this.#solution.goalBucket = startingBucket;
      }
      else if (secondBucket.amount === goal) {
        this.#solution.goalBucket = nonStartingBucket;
      }
    }

    this.#solution.otherBucket = this.#solution.goalBucket === startingBucket
      ? secondBucket.amount
      : firstBucket.amount;
  }

  solve() {
    return this.#solution;
  }
}
